docs/fomega/lazy-machine/cekmachine.tex

```diff
@@ -0,0 +1,99 @@
+\begin{array}{lrcl}
+        \textrm{Stack} & s      & ::= & f^*    \\
+ \blue{       \textrm{Environment}} & \rho, \tau & ::= & [x \mapsto V]^* \\
+        \textrm{State} & \sigma & ::= & s;\rho \compute M \quad | \quad s;\rho \return V  \quad | \quad \ckerror{} \quad | \quad \square (V,\rho)
+    \end{array}
+\]
+}
+\end{subfigure}
+
+\vspace{1mm}
+\hrule
+\vspace{2mm}
+
+\begin{subfigure}[c]{\linewidth}  %%% ---------------- CEK machine frames ---------------- %%%
+{\small
+\caption{Reduction frames}
+\[
+    \begin{array}{rlr}
+       f ::= & \inInstLeftFrame{A}                     & \textrm{left instantiation}\\
+             & \inWrapRightFrame{\alpha}{A}            & \textrm{right wrap}\\
+             & \inUnwrapFrame{}                        & \textrm{unwrap}\\
+             & \inAppLeftFrame{(M,\rho)}                 & \textrm{left application}\\
+             & \inAppRightFrame{(V,\rho)}                & \textrm{right application}\\
+             & \inBuiltin{bn}{A^*}{V^*}{\_}{M^*}        & \textrm{builtin}\\
+    \end{array}
+\]
+}
+\end{subfigure}
+\vspace{1mm}
+\hrule
+\vspace{2mm}
+\begin{subfigure}[c]{\linewidth}  %%% ---------------- CEK machine transitions ---------------- %%%
+{
+\small
+\caption{Transitions}
+    \begin{alignat*}{2}
+     \blue{s; \rho} &\bcompute \blue{x}          &{}\bmapsto{}&\blue{s;\tau \return V \quad (\rho[x] = (V,\tau))}\\  % New rule
+% Continuing with tau worries me
+      s; \rho &\compute \con{cn}                 &{}\mapsto{}& s;\rho \return \con{cn}\\
+      s; \rho &\compute \abs{\alpha}{K}{V}       &{}\mapsto{}& s;\rho \return \abs{\alpha}{K}{V}\\
+      s; \rho &\compute \inst{M}{A}              &{}\mapsto{}& s,\inInstLeftFrame{A};\rho \compute M\\
+      s; \rho &\compute \wrap{\alpha}{A}{M}      &{}\mapsto{}& s,\inWrapRightFrame{\alpha}{A};\rho  \compute M\\ 
+      s; \rho &\compute \unwrap{M}               &{}\mapsto{}& s,\inUnwrapFrame{};\rho  \compute M\\
+      s; \rho &\compute \lam{x}{A}{M}            &{}\mapsto{}& s;\rho \return \lam{x}{A}{M}\\
+      s; \rho &\compute \app{M}{N}               &{}\mapsto{}& s,\inAppLeftFrame{(N,\rho)};\rho \compute M\\
+      s; \rho &\compute \builtin{bn}{\repetition{A}}{} &{}\mapsto{}& U \quad (\textit{$bn$ computes on $\repetition{A}$ to $U$})\\
+      s; \rho &\compute \builtin{bn}{\repetition{A}}{M \repetition{M}} &{}\mapsto{}& s,\inBuiltin{bn}{\repetition{A}}{}{\_}{\repetition{M}};\rho \compute M\\
+      s; \rho &\compute \error{A} &{}\mapsto{}& \ckerror{}\\
+      \cdot; \rho &\return V &{}\mapsto{}& \square (V, \rho)\\
+      s,\inInstLeftFrame{A}; \rho &\return \abs{\alpha}{K}{M} &{}\mapsto{}& s;\rho \compute{M} \\
+      s,\inWrapRightFrame{\alpha}{A}; \rho &\return V &{}\mapsto{}& s;\rho \return \wrap{\alpha}{A}{V}\\
+      s,\inUnwrapFrame{}; \rho &\return \wrap{\alpha}{A}{V} &{}\mapsto{}& s;\rho \return V\\
+      s,\inAppLeftFrame{(N,\rho^{\prime})}; \rho &\return V &{}\mapsto{}& s, \inAppRightFrame{(V,\rho)};\rho^{\prime} \compute N\\
```

Comment:
This seems wrong - it doesn't use `rho'` and you're only putting a value in the environment, not a closure. I think the rhs should be `s; rho'[x - > (V, rho)`

Reply:
Ah, you're right.  Well spotted!  

... and that is in fact what the implementation does.

Reply:
That's how I worked out what it should be doing ;)